AmplifyPlugins/Core/AWSPluginsCore/Model/Support/Model+GraphQL.swift

```diff
@@ -18,76 +18,129 @@ extension Model {
+    }
+
+    /// This is to account for Models with an explicit field on the Model along with an object representing the hasOne
+    /// association to another model. See https://github.com/aws-amplify/amplify-ios/issues/920 for more details.
+    /// When the associated object is `nil`, remove the key from the GraphQL input to prevent runtime failures.
+    /// When the associated object is found, take this value over the explicit field's value by replacing the correct
+    /// entry for the field name of the associated model.
+    private func fixHasOneAssociationsWithExplicitFieldOnModel(_ input: GraphQLInput,
+                                                               modelSchema: ModelSchema) -> GraphQLInput {
+        var input = input
+        modelSchema.fields.forEach {
+            let modelField = $0.value
+            if case .model = modelField.type,
+               case .hasOne = modelField.association,
+               input.keys.contains(modelField.name) {
+
+                let modelIdOrNilOptional = input[modelField.name]
+                // swiftlint:disable:next syntactic_sugar
+                guard case .some(Optional<Any>.some(let modelIdOrNil)) = modelIdOrNilOptional else {
+                    input.removeValue(forKey: modelField.name)
+                    return
+                }
+                if let modelIdValue = modelIdOrNil as? String {
+                    let fieldName = getFieldNameForAssociatedModels(modelField: modelField)
+                    input[fieldName] = modelIdValue
+                }
+            }
+        }
+
         return input
     }
+
+    private func getModelId(from value: Any, modelSchema: ModelSchema) -> String {
+        if let modelValue = value as? Model {
+            return modelValue.id
+        } else if let value = value as? [String: JSONValue],
+                  case .string(let primaryKeyValue) = value[modelSchema.primaryKey.name] {
+            return primaryKeyValue
+        }
+
+        preconditionFailure("Could not retrieve the identifier from associated model: \(value)")
+    }
+
+    private func getFieldValue(for modelFieldName: String, modelSchema: ModelSchema) -> Any?? {
+        if let jsonModel = self as? JSONValueHolder {
+            return jsonModel.jsonValue(for: modelFieldName, modelSchema: modelSchema) ?? nil
+        } else {
+            return self[modelFieldName] ?? nil
+        }
+    }
+
+    /// Retrieves the GraphQL field name that associates the current model with the target model.
+    /// By default, this is the current model + the associated Model + "Id", For example "comment" + "Post" + "Id"
+    /// This information is also stored in the schema as `targetName` which is codegenerated to be the same as the
+    /// default or an explicit field specified by the developer.
+    private func getFieldNameForAssociatedModels(modelField: ModelField) -> String {
+        let defaultFieldName = modelName.camelCased() + modelField.name + "Id"
```

Comment:
"Id" was not added before this change.. is this a bug fix ? Earlier it was `modelName.camelCased() + field.graphQLName`

Reply:
`graphQLName` internally resolves to `name.pascalCased() + "Id"` for associations.  oh shoot. this is incorrect then it should be:
`modelName.camelCased() + modelField.name.pascalCased() + "Id"`

Reply:
updated to correct association field, we always pascal case the `modeField.name` for the associated field

Reply:
Cool! Wondering whether we have coverage on this logic. It would nice if you can add this or add a backlog to add this logic to unit test.